external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala

```diff
@@ -179,6 +179,56 @@ private[kafka010] class KafkaOffsetReader(
     KafkaSourceOffset(fetched)
   }
 
+  def fetchSpecificTimestampBasedOffsets(topicTimestamps: Map[String, Long]): KafkaSourceOffset = {
+    val fetched = runUninterruptibly {
+      withRetriesWithoutInterrupt {
+        // Poll to get the latest assigned partitions
+        consumer.poll(0)
+        val partitions = consumer.assignment()
+
+        // Call `position` to wait until the potential offset request triggered by `poll(0)` is
+        // done. This is a workaround for KAFKA-7703, which an async `seekToBeginning` triggered by
+        // `poll(0)` may reset offsets that should have been set by another request.
+        partitions.asScala.map(p => p -> consumer.position(p)).foreach(_ => {})
+
+        consumer.pause(partitions)
+
+        val assignedTopics = partitions.asScala.map(_.topic())
+        assert(assignedTopics == topicTimestamps.keySet,
+          "If starting/endingOffsetsByTimestamp contains specific offsets, you must specify all " +
+            s"topics. Specified: ${topicTimestamps.keySet} Assigned: $assignedTopics")
```

Comment:
We should test that 0 and System.currentTimeMillis() work straightforwardly as timestamp offsets, so people can use them the same way as -2 and -1 in the non-timestamp case.

Reply:
Given we want to state that Spark transparently passes timestamp to Kafka without interpreting and reasoning, would we need to add some test to reason about timestamp semantics?

Reply:
I'm never opposed to more tests :)